WDE.SmartScriptEditor/History/SmartDataGroupsHistory.cs

```diff
@@ -0,0 +1,223 @@
+            foreach (var data in groupsData)
+                UnbindGroupData(data);
+        }
+
+        private void OnDataCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
+        {
+            if (e.Action == NotifyCollectionChangedAction.Add)
+            {
+                if (e.NewItems != null)
+                {
+                    foreach(var item in e.NewItems)
+                        PushAction(new SmartDataGroupHistoryAction(item as SmartDataGroupsEditorData, e.NewStartingIndex,
+                            GroupHistoryActionMode.ACTION_ADD, groupsData));
+                }
+            }
+            else if (e.Action == NotifyCollectionChangedAction.Remove)
+            {
+                if (e.OldItems != null)
+                {
+                    foreach(var item in e.OldItems)
+                        PushAction(new SmartDataGroupHistoryAction(item as SmartDataGroupsEditorData, e.OldStartingIndex,
+                            GroupHistoryActionMode.ACTION_REMOVE, groupsData));
+                }
+            }
+        }
+
+        private void OnDataNameChanged(SmartDataGroupsEditorData data, string newName, string oldName)
+        {
+            PushAction(new SmartDataGroupNameHistoryModifedAction(data, oldName, newName));
+        }
+
+        private void OnGroupMembersCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
+        {
+            if (e.Action == NotifyCollectionChangedAction.Add)
+            {
+                if (e.NewItems != null)
+                {
+                    foreach (var item in e.NewItems)
+                    {
+                        var node = item as SmartDataGroupsEditorDataNode;
+                        PushAction(new SmartDataGroupMemberHistoryAction(node, e.NewStartingIndex,
+                            GroupHistoryActionMode.ACTION_ADD, node.Owner, node.Owner.Members));
+                    }
+                }
+            }
+            else if (e.Action == NotifyCollectionChangedAction.Remove)
+            {
+                if (e.OldItems != null)
+                {
+                    foreach (var item in e.OldItems)
+                    {
+                        var node = item as SmartDataGroupsEditorDataNode;
+                        PushAction(new SmartDataGroupMemberHistoryAction(node, e.OldStartingIndex,
+                            GroupHistoryActionMode.ACTION_REMOVE, node.Owner, node.Owner.Members));
+                    }
+                }
+            }
+        }
+
+        private void BindGroupData(SmartDataGroupsEditorData data)
+        {
+            data.OnNameChanged += OnDataNameChanged;
+            data.Members.CollectionChanged += OnGroupMembersCollectionChanged;
+        }
+
+        private void UnbindGroupData(SmartDataGroupsEditorData data)
+        {
+            data.OnNameChanged -= OnDataNameChanged;
+            data.Members.CollectionChanged -= OnGroupMembersCollectionChanged;
+        }
+    }
```

Comment:
Replace binding to groupsData with this:

```
groupsData.ToStream().Subscribe(e =>
            {
                if (e.Type == CollectionEventType.Add)
                {
                    PushAction(new SmartDataGroupHistoryAction(e.Item, e.Index,
                        GroupHistoryActionMode.ACTION_ADD, groupsData));
                    BindGroupData(e.Item);
                } 
                else if (e.Type == CollectionEventType.Remove)
                {
                    UnbindGroupData(e.Item);
                    PushAction(new SmartDataGroupHistoryAction(e.Item, e.Index,
                        GroupHistoryActionMode.ACTION_REMOVE, groupsData));
                }
            });
```
(remember to dispose this subscription). This solves problem mentioned above and makes code a bit shorter.